File: 0912-sort-an-array/0912-sort-an-array.py
class Solution:
    
    def sortArray(self, nums: List[int]) -> List[int]:
        def divide(l : int , r : int) : 
            if l >= r : 
                return 
            mid = (l + r) // 2
            divide(l, mid)
            divide(mid+1,r)
            merge(l, mid, r)
        
        def merge(l : int , mid : int, r : int) : 
            s1, s2, n1, n2 = l, mid+1, mid - l + 1, r - mid
            for i in range(n1) : 
                temp[s1 + i] = nums[s1 + i]
            for i in range(n2) : 
                temp[s2 + i] = nums[s2 + i]
            
            i , j , k = 0 , 0, l
            while i < n1 and j < n2 : 
                if temp[s1 + i] <= temp[s2 + j] : 
                    nums[k] = temp[s1 + i]
                    i+= 1
                else : 
                    nums[k] = temp[s2 + j]
                    j+= 1
                k += 1
            while i < n1 : 
                nums[k] = temp[s1 + i]
                k+=1
                i+=1
            while j < n2 : 
                nums[k] = temp[s2 + j]
                k+=1
                j+=1
        temp = [0] * len(nums)  
        divide(0, len(nums) -1)
        return nums
        

File: 0912-sort-an-array/0912-sort-an-array.cpp
class Solution {
public:
    void merge(vector<int>& num, int l ,int mid, int r){
        vector<int> ar1, ar2, ar3;
        ar1.insert(ar1.end(), num.begin() + l, num.begin() + mid + 1);
        ar2.insert(ar2.end(), num.begin() + mid + 1, num.begin() + r + 1); 
        int n = mid - l + 1;
        int m = r - mid;
        int i = 0, j = 0 , k = l;
        while(i < n && j < m){
            if(ar1[i] <= ar2[j]) num[k++] = ar1[i++];
            else num[k++] = ar2[j++];
        }
        while(i < n) num[k++] = ar1[i++];
        while(j < m) num[k++] = ar2[j++];
    }
    void divide(vector<int>& num, int l, int r){
        if(l >= r) return;
        int mid = l + (r-l)/2;
        divide(num, l, mid);
        divide(num, mid+1, r);
        merge(num, l ,mid, r);
    }
    vector<int> sortArray(vector<int>& nums) {
        divide(nums, 0, nums.size() - 1);
        return nums;
    }
};